R/data.table.R

```diff
@@ -1861,90 +1861,243 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
           warning("rownames is TRUE but key has multiple columns ",
                   brackify(key(x)), "; taking first column x[,1] as rownames")
         }
-        rownames = if (length(key(x))==1L) chmatch(key(x),names(x)) else 1L
+        rownames.index = if (length(key(x))==1L) chmatch(key(x),names(x)) else 1L
       }
       else if (is.logical(rownames) || is.na(rownames)) {
         # FALSE, NA, NA_character_ all mean the same as NULL
         rownames = NULL
       }
       else if (is.character(rownames)) {
-        w = chmatch(rownames, names(x))
-        if (is.na(w)) stop("'", rownames, "' is not a column of x")
-        rownames = w
+        rownames.index = chmatch(rownames, names(x))
+        if (is.na(rownames.index)) stop("'", rownames, "' is not a column of x")
+        rownames = paste0("'", rownames, "'") # for use in later error messages if needed
       }
       else { # rownames is a column number already
-        rownames = as.integer(rownames)
-        if (is.na(rownames) || rownames<1L || rownames>ncol(x))
+        rownames.index = as.integer(rownames)
+        if (is.na(rownames.index) || rownames.index<1L || rownames.index>ncol(x))
           stop("as.integer(rownames)==", rownames,
                " which is outside the column number range [1,ncol=", ncol(x), "].")
       }
     }
+    if (!is.null(rownames)) {
+      # Extract the rownames column, bringing it into memory if ff object
+      if (is.ff(x[[rownames.index]])) { 
+        rownames.value = x[[rownames.index]][]
```

Comment:
this should probably be `nocov`, we can probive an artificial test but nocov seems more appropriate